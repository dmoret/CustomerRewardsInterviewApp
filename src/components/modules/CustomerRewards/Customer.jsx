/**
 * Customer Rewards
 * displays customer record
 * @version 0.1
 * @author Daniel Moret
 */

import React from "react";
import { BrowserRouter as Router, Link } from "react-router-dom";

export default function Customer({ customer }) {
  return (
    <tr>
      <td className="customer-id">{customer.id}</td>
      <td className="customer-name">
        {customer.lastName}, {customer.firstName}
      </td>
      <td className="total-spent">${customer.totalSpent}</td>
      <td className="reward-points">{customer.rewardPoints}</td>
      <td className="created-at">{customer.createdAt}</td>
      <td className="updated-at">{customer.updatedAt}</td>
      <td>
        <Router>
          <Link to="#">View Orders</Link>
        </Router>
      </td>
    </tr>
  );
}
