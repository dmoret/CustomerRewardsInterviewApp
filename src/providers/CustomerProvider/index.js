/**
 * Customer Provider
 * @version 0.1
 * @author  Daniel Moret
 */

import React, { useState, useEffect } from "react";
import { handleResponse } from "api/methods";
import { CustomerContext } from "./context";
export * from "./context";

export const rewardPriceFirstTier = 50;
export const rewardPriceSecondTier = 100;

export const CustomerProvider = (props) => {
  const { children } = props;
  const [customerId, setCustomerId] = useState("");
  const [customers, setCustomers] = useState([]);
  const [customerRewards, setCustomerRewards] = useState([]);

  /**
   * Get all customers
   * mock up API call, no auth set
   * @param   id  integer  Optional. Customer ID
   * @returns res array    All or single customer
   */
  const getCustomers = async (id = "") => {
    console.log("CustomerProvider ::  getCustomers .... ");
    return fetch(`${process.env.REACT_APP_API_URL}/customers/${id}`).then(async (response) => {
      let res = await handleResponse(response);
      res = Array.isArray(res) ? res : [res];
      setCustomers(res);
    });
  };

  /**
   * Set Customer ID
   * convert customer ID string to number and set
   * @param customerId string  Required. The customer ID
   */
  const setCustomerById = (customerId) => {
    customerId = parseInt(customerId);
    isNaN(customerId) ? setCustomerId("") : setCustomerId(customerId);
  };

  const value = {
    state: {
      customerId,
      customers,
      customerRewards,
    },
    actions: {
      getCustomers,
      setCustomerById,
    },
  };

  // Get and calculate customer rewards from loaded customers
  useEffect(() => {
    if (customers.length) {
      const getCustomerRewards = () => {
        console.log("CustomerProvider :: getCustomerRewards :: customers :: ", customers);
        if (customers.length) {
          const customerRewardsResult = customers.map((customer) => {
            const totalSpent = customer.totalSpent / 100;
            let rewardCalcAmount = Math.floor(totalSpent);
            const totalSpentFloor = rewardCalcAmount;

            // Reward 1 point for every dollar if customer spent over fist tier amount
            if (rewardCalcAmount > rewardPriceFirstTier) {
              rewardCalcAmount = rewardCalcAmount - 50;
              // Get amount to reward if price is over second tier amount
              if (totalSpentFloor > rewardPriceSecondTier) {
                const amount = totalSpentFloor - rewardPriceSecondTier;
                rewardCalcAmount = rewardCalcAmount - amount;
              }
              customer.rewardPoints += 1 * rewardCalcAmount;
            }

            // Reward 2 points for every dollar if customer spent over $100
            if (totalSpentFloor > rewardPriceSecondTier) {
              const amount = totalSpentFloor - rewardPriceSecondTier;
              customer.rewardPoints += amount * 2;
            }

            // Format customer data for display
            customer.totalSpent = totalSpent.toFixed(2);
            // I usually use moment.js for these calculations, but the assesment asked to avoid using unnecessary 3rd party libraries
            // I do feel its necessary for scalability, but haven't included it since this app its for an assesement test by interviewers.
            customer.createdAt = new Date(customer.createdAt * 1000).toLocaleDateString("en-US", {
              day: "numeric",
              month: "short",
              year: "numeric",
            });
            customer.updatedAt = new Date(customer.updatedAt * 1000).toLocaleDateString("en-US", {
              day: "numeric",
              month: "short",
              year: "numeric",
            });

            return customer;
          });

          setCustomerRewards(customerRewardsResult);
        }
      };
      getCustomerRewards();
    }
  }, [customers]);

  // Get all customers or single customer by ID
  useEffect(() => {
    (async () => {
      //await getCustomers(customerId);
      console.debug("parseInt(customerId) :: ", parseInt(customerId));
      if (parseInt(customerId)) {
        console.debug("CustomerProvider :: useEffect :: state.customerId :: ", customerId);
        await getCustomers(customerId);
      } else {
        console.debug("CustomerProvider :: useEffect :: state.customerId :: ", customerId);
        await getCustomers();
      }
    })();
  }, [customerId]);

  return <CustomerContext.Provider value={value}>{children}</CustomerContext.Provider>;
};
