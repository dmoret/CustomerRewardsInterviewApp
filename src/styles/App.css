/**
* App Stylesheet
* @version 0.1
* @author  Daniel Moret
*/

@import "global/global.css";
@import "global/tables.css";
@import "global/forms.css";
@import "global/helpers.css";
