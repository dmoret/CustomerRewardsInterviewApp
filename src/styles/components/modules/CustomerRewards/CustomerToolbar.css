/*  */
/**
* CustomerToolbar Stylesheet
* @version 0.1
* @author  Daniel Moret
*/
#customer-toolbar-form {
  margin-bottom: 1rem;
}
#customer-toolbar-form label {
  margin: 0 0.5rem;
}
