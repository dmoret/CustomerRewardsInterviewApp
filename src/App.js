/**
 * Customer Rewards App
 * @version 0.1
 * @author Daniel Moret
 */

import React from "react";
import { CustomerProvider } from "providers/CustomerProvider";
import CustomerRewards from "components/modules/CustomerRewards";
import "styles/App.css";

export default function App() {
  return (
    <div className="App">
      <CustomerProvider>
        <CustomerRewards />
      </CustomerProvider>
    </div>
  );
}
