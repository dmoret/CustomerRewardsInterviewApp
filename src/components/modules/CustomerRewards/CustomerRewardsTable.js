/**
 * Customer Rewards Table
 * displays customer record
 * @version 0.1
 * @author Daniel Moret
 */

import React from "react";
import Customer from "./Customer";
import { useCustomerContext } from "providers/CustomerProvider";

export default function CustomerRewardsTable() {
  const { state } = useCustomerContext();
  return (
    <div>
      <table id="table-customer-rewards">
        <thead>
          <tr>
            <th>ID</th>
            <th>Name</th>
            <th>Total Spent</th>
            <th>Reward Points</th>
            <th>Date Created</th>
            <th>Date Updated</th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          {state.customerRewards.length &&
            state.customerRewards.map((customer) => (
              <Customer key={customer.id} customer={customer} />
            ))}
        </tbody>
      </table>
    </div>
  );
}
