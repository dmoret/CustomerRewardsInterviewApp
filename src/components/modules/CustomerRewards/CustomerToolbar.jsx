/**
 * Customer Rewards Table
 * displays customer record
 * @version 0.1
 * @author Daniel Moret
 */

import React from "react";
import { useCustomerContext } from "providers/CustomerProvider";

export default function CustomerToolbar() {
  const { state, actions } = useCustomerContext();

  const onInputChange = async (customerId) => {
    actions.setCustomerById(customerId);
  };

  const onHandleFocus = (event) => event.target.select();

  return (
    <div className="customer-toolbar">
      <form>
        <label>
          Customer ID#:
          <input
            type="text"
            value={state.customerId}
            onInput={(event) => onInputChange(event.target.value)}
            onFocus={onHandleFocus}
          />
        </label>
      </form>
    </div>
  );
}
