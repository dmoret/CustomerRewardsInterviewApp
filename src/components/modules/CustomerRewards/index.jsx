/**
 * Customer Rewards
 * calculates and displays customer rewards
 * @version 0.1
 * @author Daniel Moret
 */

import React from "react";
import { useCustomerContext } from "providers/CustomerProvider";
import CustomerRewardsTable from "components/modules/CustomerRewards/CustomerRewardsTable";
import LoadingPlaceholder from "components/common/LoadingPlaceholder";
import CustomerToolbar from "./CustomerToolbar";

export default function CustomerRewards() {
  const { state } = useCustomerContext();
  return (
    <div className="container">
      <h2>Customer Rewards</h2>
      <CustomerToolbar />
      {state.customerRewards.length ? <CustomerRewardsTable /> : <LoadingPlaceholder />}
    </div>
  );
}
