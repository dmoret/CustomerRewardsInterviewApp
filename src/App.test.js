/**
 * Customer Rewards App Testing
 * @version 0.1
 * @author Daniel Moret
 *
 * Testing priorities
 *
 * 1. High value features
 * 2. Edge cases in high value features
 * 3. Things that are easy to break
 * 4. Basic React component testing
 *    - User interactions
 *    - Conditional rendering
 *    - Utils / hooks
 */

import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import App from "./App";
import { CustomerProvider } from "providers/CustomerProvider";
import CustomerRewards from "components/modules/CustomerRewards";
import CustomerToolbar from "components/modules/CustomerRewards/CustomerToolbar";

const setup = () => {
  const utils = render(<CustomerToolbar />);
  const input = utils.getByLabelText("Customer ID#:");
  return {
    input,
    ...utils,
  };
};

describe("Loads Compenent :: App", () => {
  test("renders App component", () => {
    render(<App />);
    screen.debug();
  });
});

describe("Loads Provider :: CustomerProvider", () => {
  test("renders CustomerProvider", () => {
    render(<CustomerProvider />);
  });
});

describe("Loads Compenent :: CustomerRewards", () => {
  test("renders CustomerProvider component", () => {
    render(<CustomerRewards />);
    screen.debug();
  });
});

describe("Inital Load :: Compenent :: CustomerRewards", () => {
  test("renders CustomerRewards component", () => {
    render(<App />);

    // Check for title customer rewards
    expect(screen.getByText(/customer rewards/i)).toBeInTheDocument();

    // Check for table loader
    expect(screen.getByText(/loading/i)).toBeInTheDocument();

    screen.debug();
  });
});

describe("CustomerToolbar :: Input Field :: Customer ID", () => {
  test("It should not allow letters to be inputted", () => {
    const { input } = setup();
    expect(input.value).toBe(""); // empty before
    fireEvent.change(input, { target: { value: "asd" } });
    expect(input.value).toBe(""); //empty after
  });
});

describe("Compenent :: CustomerRewards :: Get all and single customer", () => {
  test("renders a single customer with rewards into table", async () => {
    await act(async () => {
      render(
        <App>
          <CustomerProvider>
            <CustomerRewards />
          </CustomerProvider>
        </App>
      );
    });

    const input = screen.getByLabelText("Customer ID#:");

    await waitFor(() => {
      // Check if there are any customers loading
      expect(screen.getAllByRole("row").length).toBeGreaterThan(1);
      // Check for 'View Order' button
      expect(screen.getAllByText("View Orders")[0]).toBeInTheDocument();
      //screen.debug();
    });

    await waitFor(() => {
      // Input customer ID to retrieve a single customer
      fireEvent.input(input, { target: { value: "2" } });
      expect(screen.getAllByRole("row").length).toEqual(2);
    });
  });
});

describe("Matches snapshot", () => {
  test("match the snapshot", () => {
    const { asFragment } = render(
      <App>
        <CustomerProvider>
          <CustomerRewards />
        </CustomerProvider>
      </App>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
