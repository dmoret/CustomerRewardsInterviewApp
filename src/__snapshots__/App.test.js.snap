// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="container"
    >
      <h2>
        Customer Rewards
      </h2>
      <div
        class="customer-toolbar"
      >
        <form>
          <label>
            Customer ID#:
            <input
              type="text"
              value=""
            />
          </label>
        </form>
      </div>
      <div>
        <p>
          Loading...
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
